### PR TITLE
Refactor export map cache to not store transient symbols

### PR DESCRIPTION
**Pull request by andrewbranch**
Follow-up to #44176 in a way, but fixes a memory leak that has been present for much longer. Previously, the cache stored exported symbols as-is, including transient ones, which can store a Type, which stores the TypeChecker it came from. We were never retaining more than one stale checker at a time, and were clearing it frequently as the program changed shape and the user moved between different files. Still, it was a mistake, and also caused part of #44176 to not work in certain edge cases (see the test added to `completionsIncomplete.ts`).

This PR also renames and relocates a lot of things to make more sense. I’ll point out where the significant changes are in review comments.

### CODE REVIEW

File: src/services/exportInfoMap.ts [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+                    symbolName: importedName,
+                    moduleName,
+                    moduleFile,
+                    moduleFileName: moduleFile?.fileName,
+                    exportKind,
+                    isTypeOnly: !(skipAlias(symbol, checker).flags & SymbolFlags.Value),
+                    isFromPackageJson,
+                    symbol: storedSymbol,
+                    moduleSymbol: storedModuleSymbol,
+                });
```
Comment (andrewbranch) [inline src/services/exportInfoMap.ts:99]: Symbols stored in `symbols` if transient, `exportInfo` if not. I tried never storing _any_ symbols, but the performance hit was too great.

File: src/server/project.ts [hunk 2]
```diff
@@ -251,7 +251,7 @@ namespace ts.server {
         public readonly getCanonicalFileName: GetCanonicalFileName;
 
         /*@internal*/
-        private exportMapCache = createExportMapCache();
+        exportMapCache: ExportInfoMap | undefined;
```
Comment (amcasey) [inline src/server/project.ts:254]: It seems weird to expose this directly and via `getCachedExportInfoMap`.  It might be cleaner to add `clearCachedExportInfoMap` instead.

File: src/services/exportInfoMap.ts [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        // which will invalidate the export map cache if things change, so pull it before
+        // checking the cache.
+        host.getPackageJsonAutoImportProvider?.();
+        const cache = host.getCachedExportInfoMap?.() || createCacheableExportInfoMap({
+            getCurrentProgram: () => program,
+            getPackageJsonAutoImportProvider: () => host.getPackageJsonAutoImportProvider?.(),
+        });
+
+        if (cache.isUsableByFile(importingFile.path)) {
+            host.log?.("getSymbolToExportInfoMap: cache hit");
```
Comment (amcasey) [inline src/services/exportInfoMap.ts:319]: "getExportInfoMap"?

File: src/services/completions.ts [hunk 4]
```diff
@@ -1976,6 +1976,7 @@ namespace ts.Completions {
             }
         }
 
+
```
Comment (amcasey) [inline src/services/completions.ts:1979]: This looks unintentional.

File: src/services/completions.ts [hunk 5]
```diff
@@ -1925,28 +1925,28 @@ namespace ts.Completions {
...
                 !!importCompletionNode,
                 context => {
-                    exportInfo.forEach(getChecker, (info, symbolName, isFromAmbientModule) => {
+                    exportInfo.forEach(sourceFile.path, (info, symbolName, isFromAmbientModule) => {
                         if (!detailsEntryId && isStringANonContextualKeyword(symbolName)) return;
                         const isCompletionDetailsMatch = detailsEntryId && some(info, i => detailsEntryId.source === stripQuotes(i.moduleSymbol.name));
                         if (isCompletionDetailsMatch || charactersFuzzyMatchInString(symbolName, lowerCaseTokenText)) {
-                            if (isFromAmbientModule && !some(info, isImportableExportInfo)) {
+                            const defaultExportInfo = find(info, isImportableExportInfo);
+                            if (isFromAmbientModule && !defaultExportInfo) {
```
Comment (amcasey) [inline src/services/completions.ts:1933]: Are the two conditions related?  If not, how do we know `exportInfo` is defined below?

File: src/services/exportInfoMap.ts [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+    export interface CacheableExportInfoMapHost {
+        getCurrentProgram(): Program | undefined;
+        getPackageJsonAutoImportProvider(): Program | undefined;
+    }
+
+    export function createCacheableExportInfoMap(host: CacheableExportInfoMapHost): ExportInfoMap {
+        let exportInfoId = 1;
+        const exportInfo = createMultiMap<string, CachedSymbolExportInfo>();
+        const symbols = new Map<number, [symbol: Symbol, moduleSymbol: Symbol]>();
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/exportInfoMap.ts:65]: This isn't stored as an array because we want the id to persist across clears?

File: src/testRunner/unittests/tsserver/exportMapCache.ts [hunk 7]
```diff
@@ -23,54 +23,109 @@ namespace ts.projectSystem {
...
+            declare const exp: {} & { [K in Signals]: K };
+            export = exp;`,
+    };
 
     describe("unittests:: tsserver:: exportMapCache", () => {
         it("caches auto-imports in the same file", () => {
-            const { exportMapCache, checker } = setup();
-            assert.ok(exportMapCache.get(bTs.path as Path, checker));
+            const { exportMapCache } = setup();
+            assert.ok(exportMapCache.isUsableByFile(bTs.path as Path));
```
Comment (amcasey) [inline src/testRunner/unittests/tsserver/exportMapCache.ts:37]: Probably out of scope, but I would find `isTrue` and `isFalse` more readable.

File: src/testRunner/unittests/tsserver/exportMapCache.ts [hunk 8]
```diff
@@ -23,54 +23,109 @@ namespace ts.projectSystem {
...
             host.runQueuedTimeoutCallbacks();
             project.getPackageJsonAutoImportProvider();
-            assert.isUndefined(exportMapCache.get(bTs.path as Path, checker));
+            assert.ok(!exportMapCache.isUsableByFile(bTs.path as Path));
+            assert.ok(exportMapCache.isEmpty());
+        });
+
+        it("does not store transient symbols through program updates", () => {
+            const { exportMapCache, project, session } = setup();
+            // SIGINT, exported from /lib/foo/constants.d.ts, is a mapped type property, which will be a trasient symbol.
```
Comment (amcasey) [inline src/testRunner/unittests/tsserver/exportMapCache.ts:78]: Typo: "trasient"

File: src/testRunner/unittests/tsserver/exportMapCache.ts [hunk 9]
```diff
@@ -23,54 +23,109 @@ namespace ts.projectSystem {
...
+                        textChanges: [{
+                            newText: " ",
+                            start: { line: 1, offset: 1 },
+                            end: { line: 1, offset: 1 },
+                        }]
+                    }]
+                }
+            });
+            project.getLanguageService(/*ensureSynchronized*/ true);
+            assert(programBefore !== project.getCurrentProgram()!);
```
Comment (amcasey) [inline src/testRunner/unittests/tsserver/exportMapCache.ts:108]: `assert.notEqual`?

**Review by amcasey — APPROVED** [review]
Seems reasonable.

File: src/services/exportInfoMap.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/exportInfoMap.ts:65]: Right, and subsequent cache sets could happen in essentially random order, and I believe that Maps are better-performing than sparse arrays.

File: src/services/completions.ts (hunk 5 above)
Comment (andrewbranch) [reply src/services/completions.ts]: I made a mental note to revisit this and see if I could break these conditions apart, but I forgot to at the time. It now makes more sense.
